### ISSUE TITLE
Type of a character literal is determined by the value not the real type

### ISSUE THREAD

**Issue by paxbun** [reitermarkus/cmacro-rs#3]
https://github.com/reitermarkus/cmacro-rs/blob/2e47e1bc4544d5af106d52b66225ea5fa13cc94a/src/ast/literal.rs#L302-L325

It seems like `cmacro-rs` converts all character literals into `u32` values and determines the type of those literals by their value. For example, `cmacro-rs` thinks `CHAR_32` in the following code is a `u16` value.

```cpp
#define CHAR_16 u'\uAC00'
#define CHAR_32 U'\uAC00'
```

I think `cmacro-rs` should preserve prefix information in `literal::LitChar`, and use it to determine the correct type as follows:
```rust
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LitChar {
  pub(crate) repr: u32,
  pub(crate) prefix: LitCharPrefix
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LitCharPrefix {
  Char, // no prefix
  WChar, // L
  Char8, // u8
  Char16, // u
  Char32, // U
}
```

This can be used to support Unicode character literals as well. rust-lang/rust-bindgen#2369 ignores Unicode character literals representing codepoints greater than 256, because of the following implementation.
https://github.com/reitermarkus/cmacro-rs/blob/2e47e1bc4544d5af106d52b66225ea5fa13cc94a/src/ast/literal.rs#L290-L300

We can modify that part as follows.
```rust
#[allow(unused_variables)]
pub(crate) fn finish<C>(&mut self, ctx: &mut LocalContext<'_, C>) -> Result<Option<Type>, crate::Error>
where
  C: CodegenContext,
{
  match self.prefix {
    LitCharPrefix::Char => Ok(Some(Type::BuiltIn(BuiltInType::Char))),
    LitCharPrefix::Char16 => Ok(Some(Type::BuiltIn(/* We need variants for char8_t, char16_t, and char32_t */)))
    ...
  }
}
```

**Comment by reitermarkus** [comment]
Yes, storing the prefix definitely makes sense instead of comparing to `u8::MAX`.

Probably would also make sense to convert `LitChar` into an enum.

**Comment by paxbun** [comment]
Would it be okay to add `BuiltIn::{Char8, Char16, Char32}` as well? `char{16, 32}_t` is included in C11 & C++11, and `char8_t` is (will be) included in C23 & C++20.

`char8_t`, `char16_t`, and `char32_t`'s signedness and alignment are the same as those of `unsigned char`, `uint_least16_t`, and `uint_least32_t`, so at least we need enum variants for fixed-width integers to support this.

C++11 `char16_t` and `char32_t`: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2249.html
C++20 `char8_t`: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0482r6.html

**Comment by reitermarkus** [comment]
I think we can add at least `Char16` and `Char32`.

We can also add `Char8`, but I'm not sure if we need it? As far as I understand `char8_t` depends on which C standard is used and not on a prefix.

**Comment by paxbun** [comment]
AFAIK, `u8'...'` and `char8_t` are added in C23, while in C++, `u8'...'` makes a `char` value until C++17, but `char8_t` from C++20. The only standard where `u8'...'` makes a `char` value is C++17.

Although `char8_t` has the same alignment as `unsigned char`, because `char`, `signed char`, and `unsigned char` are three different types, I think this crate should distinguish `'...'` and `u8'...'`. If one thinks this is unnecessary, they can easily ignore this with the `|` syntax:
```rust
// Assume LitChar is an enum
let lit: LitChar = ...;
match lit {
  LitChar::Char(v) | LItChar::Char8(v) => /* v is u8 */
  _ => ...
}
```

**Comment by reitermarkus** [comment]
Ah, I see. Okay, let's add `Char8` for the `u8` prefix then.
